### ISSUE TITLE
add content-sqlite option to set database location

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#1038]
As discussed in #1037, the content-sqlite database file can grow quite large and /run (a tmpfs on modern systems) is likely not the right place for it when started via systemd.  One way to relocate it would be to add a command line option to the module to allow the default database location to be overridden.

Preferably use optparse for this (see userdb module for example - there is a gotcha with position of first argument in argv)

Since we don't yet have a way to restart an instance with an existing db file, let's unconditionally clean it up and get rid of the "persist-directory" and "persist-filesystem" broker attributes.

**Comment by jameshcorbett** [comment]
Any chance this could be bumped up in priority? One of my Quartz jobs (with ~3 million job steps) recently crashed with an out-of-memory error. In my slurm.out file, I found the following messages:

```
2020-02-19T13:54:45.555903Z content-sqlite.err[0]: store: executing stmt: database or disk is full(13)
2020-02-19T13:54:45.556242Z broker.err[0]: content store: No space left on device
2020-02-19T13:54:45.556496Z content-sqlite.err[0]: store: executing stmt: database disk image is malformed(11)
2020-02-19T13:54:45.558924Z broker.err[0]: content store: Invalid argument
2020-02-19T13:54:45.556762Z content-sqlite.err[0]: store: executing stmt: database disk image is malformed(11)
2020-02-19T13:54:45.559905Z broker.err[0]: content store: Invalid argument
2020-02-19T13:54:45.559987Z content-sqlite.err[0]: store: executing stmt: database disk image is malformed(11)
2020-02-19T13:54:45.560187Z broker.err[0]: content store: Invalid argument
2020-02-19T13:54:45.560231Z content-sqlite.err[0]: store: executing stmt: database disk image is malformed(11)

---------above two lines repeat 1.5 million times-------------------

2020-02-19T14:01:32.397082Z broker.err[0]: content store: Invalid argument
2020-02-19T14:01:32.863932Z content-sqlite.err[0]: store: executing stmt: database disk image is malformed(11)
flux-job: drain: No such file or directory
slurmstepd: error: Detected 11 oom-kill event(s) in step 4602745.0 cgroup. Some of your processes may have been killed by the cgroup out-of-memory handler.
srun: error: quartz542: task 0: Out Of Memory
srun: First task exited 30s ago
srun: step:4602745.0 tasks 1-9: running
srun: step:4602745.0 task 0: exited abnormally
srun: Terminating job step 4602745.0
srun: Job step aborted: Waiting up to 62 seconds for job step to finish.
slurmstepd: error: Detected 3 oom-kill event(s) in step 4602745.0 cgroup. Some of your processes may have been killed by the cgroup out-of-memory handler.
srun: error: quartz1774: task 5: Killed
srun: error: quartz1993: task 7: Killed
srun: error: quartz1411: task 3: Killed
srun: error: quartz1545: task 4: Killed
srun: error: quartz1977: task 6: Killed
srun: error: quartz2134: task 8: Killed
srun: error: quartz999: task 2: Killed
srun: error: quartz3046: task 9: Killed
slurmstepd: error: Detected 11 oom-kill event(s) in step 4602745.batch cgroup. Some of your processes may have been killed by the cgroup out-of-memory handler.
```

After reading #609 I take it that Flux stores a SQLite db in node-local storage, which is really just memory, and hence the OOM error. So if I could redirect the SQLite database to a proper file system, I wouldn't get this error, right?

**Comment by dongahn** [comment]
Hmmm. Somehow I thought there was a way to redirect the content store to a bigger file system like scratch nfs or luster. Maybe that was only available with older versions...

**Comment by grondo** [comment]
As alluded to above, I think we still have the `persist-directory` and `persist-filesystem` broker attributes. Try adding `flux start -o,-Spersist-filesystem=/nfs/tmp2/username` or similar, that should store the sqlite db under a unique directory in that path.

Since these attributes are meant to be used to persist data, you'll have to manually remove them when your jobs are complete.

**Comment by jameshcorbett** [comment]
Got it. Thanks @grondo!

**Comment by jameshcorbett** [comment]
Do either of you know whether the attribute will be passed down to child instances? I'm using `flux tree` and I imagine it's the combination of all the instances that is causing the memory issues.

**Comment by grondo** [comment]
No, I don't think any attributes are passed down to child instances by default.

Does the current version of flux-tree support a way to set options on every instance? Unfortunately, I don't *think* that setting `persist-filesystem` attribute early in your `rc1` script will work, since `create_persistdir()` is called before runlevel 1 is entered.

**Comment by dongahn** [comment]
@jameshcorbett and @grondo: I can easily add an option to pass specific flux options all of the instances. I will propose something.

**Comment by dongahn** [comment]
@grondo and @garlick: I am working on a small addition to `flux-tree` so that it can pass `-o,-Spersist-filesystem=DIR` to each flux tree instance. 

```
ahn1@02d741c04fd5:/usr/src$ mkdir -p DIR/
ahn1@02d741c04fd5:/usr/src$ flux start -o,-Spersist-filesystem=DIR
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

ahn1@02d741c04fd5:/usr/src$ ls -l DIR
total 0
ahn1@02d741c04fd5:/usr/src$
ahn1@02d741c04fd5:/usr/src$ flux getattr persist-filesystem
DIR
```

While this sets the `persist-filesystem` correctly (per `flux-getattr`), the given directory isn't populated.

Is this expected? Maybe KVS needs to grow somewhat before `content` pushes it to `sqlite`.

Let me know. Otherwise, I am ready to do a small `flux-sched` PR.

**Comment by SteVwonder** [comment]
What about setting `rundir`?

```
❯ ./flux start -o,-Srundir=/tmp/rundir-test bash -c 'tree $(flux getattr rundir)'
/tmp/rundir-test
├── 0
│   └── local
└── content.sqlite
```

**Comment by dongahn** [comment]
> You would probably need to add logic to flux-tree to add the treeid to each rundir to make that work.

I am already doing this so this should be okay. It is just that I can't find the content. 

@grondo and @garlick:  Please confirm if `-o,-Spersist-filesystem=DIR` has been deprecated and it is now `rundir` we need to set.

**Comment by grondo** [comment]
Yes, it was dropped in 21ffb63d918403f2d8409552e0264b45e2dfb45b.

Use `rundir` or set `content.backing-path` to the path you want to place the sqlite db.

**Comment by grondo** [comment]
FYI the new systemd unit file will set the content. backing-path directly as well:
```
-Scontent.backing-path=@X_LOCALSTATEDIR@/lib/flux/content.sqlite
```

**Comment by dongahn** [comment]
Sorry. I haven't followed the new systems unit file. 

Is this a configuration file time setup? 

What I will need is a switch that allows me to override the system settings so that users can avoid memory overuse. Will `-o,-Srundir=DIR` do this job now and in the future?

**Comment by grondo** [comment]
> Is this a configuration file time setup?

This is the installed `systemd` unit file for the system instance. It will not affect any subinstances.

> What I will need is a switch that allows me to override the system settings so that users can avoid memory overuse. Will -o,-Srundir=DIR do this job now and in the future?

I think what you want is `-o,-Scontent.backing-path=`, which defines a file path for the sqlite database, and preserves the file after instance exit.

If you don't want to persist the sqlite path, but just want to keep the content store out of /tmp, then by default `content.backing-path` is set under `rundir` so you could try setting that.

From the `flux-broker-attributes(7)` mapage:
```
       rundir
           A global, shared temporary directory available for scratch storage
           within the session. By default, a temporary directory is created
           for each broker rank, but if rundir is set on the command line,
           this directory may be shared by all broker ranks (typically only
           within a node). If rundir does not exist, it is created and
           subsequently removed during session exit. An existing rundir is not
           removed at exit.

       content.backing-path
           The path to the content backing store file(s). If this is set on
           the broker command line, the backing store uses this path instead
           of a temporary one, and content is preserved on instance exit. If
           file exists, its content is imported into the instance. If it
           doesn’t exist, it is created.

```

**Comment by dongahn** [comment]
This is helpful! Thanks @grondo. `rundir` should work better in this case because users probably don't want to leave a large number of content stores in their persistent file systems after each `flux-tree` invocation.

**Comment by grondo** [comment]
Yeah, that should work well then. I think you only need a unique rundir per job, not per broker or node.

**Comment by dongahn** [comment]
> a unique rundir per job

Just to clarify: per Flux instance (i.e., per flux start)?

**Comment by grondo** [comment]
Yes, that's right.

**Comment by garlick** [comment]
I think this was resolved by PR #2783.  Please open a new issue if there are lingering problems.
